### PR TITLE
Properly resolves path_parent field queries for the @solrsearch endpoint if they are relative to the virtual root

### PR DESCRIPTION
**Pull request by elioschmutz**
The `path_parent` field query of the `@solrsearch` endpoint properly resolves paths relative to the virtual host url and joins multiple `path_parent` field queries with an OR operator.

Jira: https://4teamwork.atlassian.net/browse/CA-1588

This change is required because a frontend does not know anything about the physical path of an object which is required to do solr-path queries.

the path indexes of an object is always the physical path:

Bildschirmfoto 2021-05-27 um 13 16 22

But a frontend does not know anything about it. I.e.

the url http://www.example.com/examplecontent can point to the same physical path like the url http://localhost:8080/fd/examplecontent =>

http://www.example.com/examplecontent => `/plone/examplecontent`
http://localhost:8080/plone/examplecontent => `/plone/examplecontent`

But we can only extract the path relative to the current site domain:

http://www.example.com/examplecontent => `/examplecontent`
http://localhost:8080/plone/examplecontent => `/plone/examplecontent`

In the backend, we need to replace this path relative to the virtual host with the physical path of the content itself to be able to do a solr query.

## Checklist

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

_Only applicable should be left and checked._

- API change:
  - [ ] Documentation is updated
  - [ ] API Changelog entry (see [guide](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/451248812/API+Changelog+Guidelines))
  - If breaking:
    - [ ] api-change label added
    - [ ] #delivery channel notified about breaking change
    - [ ] Scrum master is informed
- New functionality:
  - [ ] for `document` also works for `mail`
  - [ ] for `task` also works for `forwarding`
- Further improvements needed:
  - [ ] Create follow-up stories and link them in the PR and Jira issue
- Upgrade steps (changes in profile):
  - [ ] Make it deferrable if possible
  - [ ] Execute as much as possible conditionally
- DB-Schema migration
  - [ ] All changes on a model (columns, etc) are included in a DB-schema migration.
  - [ ] Constraint names are shorter than 30 characters (`Oracle`)
- [ ] Change could impact client installations, client policies need to be adapted
- New translations
  - [ ] All msg-strings are unicode
- Change in schema definition:
  - [ ] If `missing_value` is specified, then `default` has to be set to the same value

### CODE REVIEW

File: opengever/api/solrsearch.py [hunk 1]
```diff
@@ -60,11 +61,49 @@ def extract_filters(self, params):
...
+        requested_parent_paths = []
+        portal = api.portal.get()
+        for query in list(filters):
+            if not query.startswith("path_parent:"):
+                continue
+
+            # extract the path from the query and unescape
+            path = query.split(":", 1)[1].replace('\\', '')
+
+            # A frontend does not now anything about the physical path of an object.
```
Comment (njohner) [inline opengever/api/solrsearch.py:88]: now -> know

File: opengever/api/solrsearch.py [hunk 2]
```diff
@@ -60,11 +61,49 @@ def extract_filters(self, params):
...
+            # portal path to get the relative path from the plone site root
+            relativePath = path.lstrip('/').lstrip(portal.virtual_url_path()).lstrip('/')
+
+            # And then we can extend the physical portal path with the requested
+            # relative path
+            physicalPath = portal.getPhysicalPath()
+            if relativePath:
+                physicalPath += (relativePath, )
+
+            content = portal.restrictedTraverse('/'.join(physicalPath), None)
```
Comment (njohner) [inline opengever/api/solrsearch.py:99]: I wonder whether this should not be unrestricted? I mean I should be allowed to search inside an object on which I have no view permissions no?

File: opengever/api/solrsearch.py [hunk 3]
```diff
@@ -60,11 +61,49 @@ def extract_filters(self, params):
...
+            # A frontend does not now anything about the physical path of an object.
+            # We have to take care of this. So we have to removes the virtual
+            # portal path to get the relative path from the plone site root
+            relativePath = path.lstrip('/').lstrip(portal.virtual_url_path()).lstrip('/')
+
+            # And then we can extend the physical portal path with the requested
+            # relative path
+            physicalPath = portal.getPhysicalPath()
+            if relativePath:
+                physicalPath += (relativePath, )
```
Comment (njohner) [inline opengever/api/solrsearch.py:97]: I'm surprised there is no function doing this already. But I searched and did not find anything.

File: opengever/api/solrsearch.py [hunk 4]
```diff
@@ -60,11 +61,49 @@ def extract_filters(self, params):
...
+
+            content = portal.restrictedTraverse('/'.join(physicalPath), None)
+            resolved_path = '/'.join(content.getPhysicalPath()) if content else path
```
Comment (njohner) [inline opengever/api/solrsearch.py:100]: do we need this? do you see scenarios where it does not resolve to any object but we still want to use the path as filter?

File: opengever/api/solrsearch.py (hunk 3 above)
Comment (njohner) [reply opengever/api/solrsearch.py:97]: I'm a bit confused why this is not enough. Why then resolve the object and then get the physical path from it. This is already the physical path no?

**Review by njohner — APPROVED** [review]
I'm not sure whether the part transforming the path needs to be that complicated, but otherwise LGTM

File: opengever/api/solrsearch.py (hunk 3 above)
Comment (elioschmutz) [reply opengever/api/solrsearch.py:97]: you're completely right 👍 I don't need to resolve the object

File: opengever/api/solrsearch.py (hunk 4 above)
Comment (elioschmutz) [reply opengever/api/solrsearch.py:100]: removed because there is no need to resolve the object itself.

File: opengever/api/solrsearch.py (hunk 2 above)
Comment (elioschmutz) [reply opengever/api/solrsearch.py]: removed because it's not needed anymore

**Pull request merged**